crates/core/src/client/messages.rs

```diff
@@ -22,7 +24,32 @@ pub trait ServerMessage: Sized {
     fn serialize(self, protocol: Protocol) -> DataMessage {
         match protocol {
             Protocol::Text => self.serialize_text().to_json().into(),
-            Protocol::Binary => self.serialize_binary().encode_to_vec().into(),
+            Protocol::Binary => {
+                let msg_bytes = self.serialize_binary().encode_to_vec();
+                let reader = &mut &msg_bytes[..];
+
+                // SubscriptionUpdate messages will typically be quite large,
+                // so we choose a relatively large buffer size,
+                // in this case 128KB,
+                // to optimize for compression speed.
+                // TODO(perf,bikeshedding): Measure whether allocating a large buffer here is expensive,
+                // and adjust buffer size accordingly.
+                const BUFFER_SIZE: usize = 128 * 1024;
```

Comment:
I wasn't concerned about this for SubscriptionUpdates, but this is a very large buffer for TransactionUpdates. We might need to take the hit on SubscriptionUpdates if it means not allocating for TransactionUpdates.